### ISSUE TITLE
Search throws exception - cannot connect to elastic search

### ISSUE THREAD

**Issue by TesseractG** [debiki/ed-prod-one#1]
I followed the instructions in the readme (which are really clear!) and managed to install everything without any problems.

The only issue is that the app does not seem to be able to connect to the elastic search container. Elastic search did start (docker ps shows it running), but if I look at the logs for app, it is full of the same message over and over again. If I try to search on the forum itself, I get the same issue;

```Something went wrong: [DwE500REX]

NoNodeAvailableException[None of the configured nodes are available: [{#transport#-1}{172.18.0.4}{172.18.0.4:9300}]]
	at org.elasticsearch.client.transport.TransportClientNodesService.ensureNodesAreAvailable(TransportClientNodesService.java:311)
	at org.elasticsearch.client.transport.TransportClientNodesService.execute(TransportClientNodesService.java:227)
	at org.elasticsearch.client.transport.support.TransportProxyClient.execute(TransportProxyClient.java:62)
	at org.elasticsearch.client.transport.TransportClient.doExecute(TransportClient.java:281)
	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:403)
	at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:80)
	at ed.server.search.SearchEngine.search(SearchEngine.scala:115)
	at debiki.dao.SearchDao$class.fullTextSearch(SearchDao.scala:50)
	at debiki.dao.SiteDao.fullTextSearch(SiteDao.scala:64)
	at controllers.SearchController$$anonfun$doSearch$1.apply(SearchController.scala:56)
	at controllers.SearchController$$anonfun$doSearch$1.apply(SearchController.scala:51)
	at ed.server.http.PlainApiActions$$anon$1.runBlockIfAuthOk(PlainApiActions.scala:258)
	at ed.server.http.PlainApiActions$$anon$1.invokeBlock(PlainApiActions.scala:103)
	at ed.server.http.PlainApiActions$$anon$1.invokeBlock(PlainApiActions.scala:70)
	at play.api.mvc.ActionBuilder$$anon$2.apply(Action.scala:458)
	at ed.server.http.PlainApiActions$$anon$1$$anonfun$composeAction$1.apply(PlainApiActions.scala:78)
	at ed.server.http.PlainApiActions$$anon$1$$anonfun$composeAction$1.apply(PlainApiActions.scala:77)
	at ed.server.http.SafeActions$ExceptionAction$.invokeBlock(SafeActions.scala:102)
	at ed.server.http.SafeActions$ExceptionAction$.invokeBlock(SafeActions.scala:78)
	at play.api.mvc.ActionBuilder$$anon$2.apply(Action.scala:458)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5$$anonfun$apply$6.apply(Action.scala:112)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5$$anonfun$apply$6.apply(Action.scala:112)
	at play.utils.Threads$.withContextClassLoader(Threads.scala:21)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5.apply(Action.scala:111)
	at play.api.mvc.Action$$anonfun$apply$2$$anonfun$apply$5.apply(Action.scala:110)
	at scala.Option.map(Option.scala:146)
	at play.api.mvc.Action$$anonfun$apply$2.apply(Action.scala:110)
	at play.api.mvc.Action$$anonfun$apply$2.apply(Action.scala:103)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:253)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:91)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:90)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:39)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:409)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

It seems that, for whatever reason, the app cannot connect to the elastic search back-end. This is consistent across the 2 servers where I installed the app, to make sure it was not a fluke.

**Comment by kajmagnus** [comment]
Hi, what does `sudo docker-compose logs search` show?

(For me, there's a warning: `java.lang.UnsupportedOperationException: seccomp unavailable:` but then ElasticSearch starts fine anyway, the log messages end with green status: `[INFO ][cluster.routing.allocation] [Feedback] Cluster health status changed from [RED] to [GREEN] (reason: [shards started [[all_english_v1][2]] ...])`.  In case you're wondering, I'm using ElasticSearch 5 alpha4 right now — I'm looking to upgrade to the real 5.0.0 once a Netty comatibility issue has been fixed.)

**Comment by TesseractG** [comment]
Thanks for the quick reply!

I had not yet looked at the search logs and they paint a more interesting picture. Apparently, something is running out of memory;

```
search_1  | Exception in thread "main" java.lang.RuntimeException: bootstrap checks failed
search_1  | max virtual memory areas vm.max_map_count [65530] likely too low, increase to at least [262144]
search_1  |     at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:125)
search_1  |     at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:85)
search_1  |     at org.elasticsearch.bootstrap.BootstrapCheck.check(BootstrapCheck.java:65)
search_1  |     at org.elasticsearch.bootstrap.Bootstrap$5.validateNodeBeforeAcceptingRequests(Bootstrap.java:178)
search_1  |     at org.elasticsearch.node.Node.start(Node.java:373)
search_1  |     at org.elasticsearch.bootstrap.Bootstrap.start(Bootstrap.java:193)
search_1  |     at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:252)
search_1  |     at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:96)
search_1  |     at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:91)
search_1  |     at org.elasticsearch.cli.SettingCommand.execute(SettingCommand.java:54)
search_1  |     at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:91)
search_1  |     at org.elasticsearch.cli.Command.main(Command.java:53)
search_1  |     at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:70)
search_1  |     at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:63)
search_1  | Refer to the log for complete error details.
```

docker-compose ps

```
      Name             Command             State              Ports
-------------------------------------------------------------------------
edp_app_1          /bin/sh -c rm -f   Up                 3333/tcp,
                   /opt/ed/a ...                         9000/tcp,
                                                         9443/tcp,
                                                         9999/tcp
edp_cache_1        docker-            Up                 6379/tcp
                   entrypoint.sh
                   redis ...
edp_rdb_1          /docker-           Up                 5432/tcp
                   entrypoint.sh
                   postgres
edp_search_1       /docker-           Restarting
                   entrypoint.sh
                   elas ...
edp_web_1          nginx              Up                 0.0.0.0:9443->44
                                                         3/tcp, 0.0.0.0:9
                                                         080->80/tcp
```

With ElasticSearch restarting continuously (it seems).

free -m

```
             total       used       free     shared    buffers     cached
Mem:         32105      29714       2390        954       1090      21706
-/+ buffers/cache:       6917      25188
Swap:        16375          7      16367
```

df -h

```
Filesystem      Size  Used Avail Use% Mounted on
/dev/md2        1.8T   47G  1.7T   3% /
udev             10M     0   10M   0% /dev
tmpfs           6.3G  673M  5.7G  11% /run
tmpfs            16G  8.0K   16G   1% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs            16G     0   16G   0% /sys/fs/cgroup
/dev/md1        488M   36M  427M   8% /boot
tmpfs           3.2G     0  3.2G   0% /run/user/1000
```

netstat -tlpn

```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:110             0.0.0.0:*               LISTEN      4048/dovecot
tcp        0      0 0.0.0.0:143             0.0.0.0:*               LISTEN      4048/dovecot
tcp        0      0 127.0.0.1:8080          0.0.0.0:*               LISTEN      13802/config.ru
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      766/apache2
tcp        0      0 148.251.19.49:6001      0.0.0.0:*               LISTEN      5341/laravel-echo-s
tcp        0      0 0.0.0.0:465             0.0.0.0:*               LISTEN      27501/master
tcp        0      0 127.0.0.1:8181          0.0.0.0:*               LISTEN      13769/gitlab-workho
tcp        0      0 0.0.0.0:21              0.0.0.0:*               LISTEN      13067/vsftpd
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      27515/sshd
tcp        0      0 127.0.0.1:47255         0.0.0.0:*               LISTEN      29710/redmine
tcp        0      0 127.0.0.1:5432          0.0.0.0:*               LISTEN      4928/postgres
tcp        0      0 0.0.0.0:25              0.0.0.0:*               LISTEN      27501/master
tcp        0      0 127.0.0.1:3001          0.0.0.0:*               LISTEN      24558/mongod
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      766/apache2
tcp        0      0 127.0.0.1:3306          0.0.0.0:*               LISTEN      27039/mysqld
tcp        0      0 127.0.0.1:6379          0.0.0.0:*               LISTEN      5227/redis-server 1
tcp        0      0 0.0.0.0:587             0.0.0.0:*               LISTEN      27501/master
tcp6       0      0 :::110                  :::*                    LISTEN      4048/dovecot
tcp6       0      0 :::143                  :::*                    LISTEN      4048/dovecot
tcp6       0      0 :::465                  :::*                    LISTEN      27501/master
tcp6       0      0 :::22                   :::*                    LISTEN      27515/sshd
tcp6       0      0 :::25                   :::*                    LISTEN      27501/master
tcp6       0      0 :::587                  :::*                    LISTEN      27501/master
```

scripts/stats.sh

```
CONTAINER           CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O           PIDS
edp_web_1           0.00%               0 B / 0 B           0.00%               3.21 kB / 648 B     0 B / 0 B           0
edp_app_1           80.55%              0 B / 0 B           0.00%               97.5 kB / 134 kB    0 B / 0 B           0
edp_rdb_1           0.20%               0 B / 0 B           0.00%               138 kB / 94 kB      0 B / 98.3 kB       0
edp_cache_1         0.02%               0 B / 0 B           0.00%               5.04 kB / 764 B     0 B / 0 B           0
edp_search_1        126.94%             0 B / 0 B           0.00%               0 B / 258 B         0 B / 0 B           0
```

Nothing overly special there, except the high CPU usage on both app and search perhaps. This is a pretty solid staging server that I use for all kinds of experiments. I ran elasticsearch (non-docker) before without any issues and as you can see, there is plenty of memory and disk space.

Perhaps I need some sort of special switch to assign more memory to the container?

**Comment by kajmagnus** [comment]
Thanks for the output from the commands. An error message in the output above from `docker-compose logs search` is:

> max virtual memory areas vm.max_map_count [65530] likely too low, increase to at least [262144]

I've seen that before, and this script:

https://github.com/debiki/ed-prod-one/blob/master/scripts/configure-ubuntu.sh#L24

adds this line:

> vm.max_map_count=262144    # ElasticSearch requires (at least) this, default = 65530

to `/etc/sysctl.conf`.  Did you perhaps not run the script?
(it's mentioned in step 3 here: https://github.com/debiki/ed-prod-one#installation-instructions, and ... there are many steps, I suppose it's easy to miss one step)

Would you like to run this:  `sysctl vm.max_map_count`  ?
It'll show the current setting. I think it's going to report 65530. And we need to increase it to 262144.

Since the server is being used for other stuff too, not just EffectiveDiscussions, perhaps it'd be safer if you _not_ run the `configure-ubuntu.sh` script, but instead edit `/etc/sysctl.conf` manually and add the config lines you see here, lines 22-24: https://github.com/debiki/ed-prod-one/blob/master/scripts/configure-ubuntu.sh#L22

(And if I were you, before I edit `/etc/sysctl.conf`, I would search in that file for the names of the config stuff you add, so you'll notice if you've already set those values to something else for some reason.)

and then afterwards: `sysctl --system` to reload the settings, and perhaps `sysctl vm.max_map_count` again to see if it worked.

Afterwards, you might need to do `docker-compose restart`,
or perhaps `docker-compose down` followed by `docker-compose up`

\***

(Hmm since Apache listens on port 80 and 443, I'm actually surprised you were apparently able to access EffectiveDiscussions on those ports. Did you configure ED to listen on some other port?
Or why does it work :- P  )

**Comment by TesseractG** [comment]
You are correct... I did not run that script for 2 reasons; a) I am not on Ubuntu (but on Debian) and b) this server is already in use for other stuff, so I did not want to mess up anything.

I figured all it would do was install dependencies... I should have had a look at the file! Thanks for pointing this out and my apologies for the needless ticket.

I have adapted the sysctl.conf by hand and everything runs fine now!

About the ports; I changed the docker-compose.yml to bind the ports to 9080 and 9443 (I think) on the host for testing. Once this goes into production, I will not bind the external ports at all, but instead use a reverse proxy. I generally use this package for that;

https://github.com/jwilder/nginx-proxy

Removing the ports from the docker-compose.yml and adding the environment variables makes this work pretty easily. Perhaps interesting to explore as a setup option for people who run this alongside a bunch of docker containers? - another issue entirely, though ;)

**Comment by kajmagnus** [comment]
Hi again, ok, glad it worked :-)   This was actually a helpful ticket for me: I now updated the docs to mention that there're two important lines in the `configure-ubuntu.sh` script — in case there are (for example) more people who use Debian and don't want to run the script.

If there's something else you're wondering or think I should do or change — then, you're very welcome to open a new ticket or create a topic at https://www.effectivediscussions.org/forum/.

I had a quick look at nginx-proxy, and I'm getting the impression that, in docker-compose.yml, one would remove these two ports:

```
    ports:
      # Don't expose port 81 (for publishing WebSocket events)  — it should be reachable only
      # from inside the Docker network.
      - '80:80'
      - '443:443'
```

and add this:

```
    environment:
      CDN_PULL_KEY: "public"
      VIRTUAL_HOST: effectivediscussions.example.com   # <——
```

and then start the nginx-proxy container. But not via the same docker-compose.yml file. Instead, just with the plain `docker` command (or ?).

If more people seem to be interested in this, I could add a section in the docs about how to combine ED with nginx-proxy.

**Comment by TesseractG** [comment]
It would indeed work like that, though there is no reason to start nginx-proxy outside docker-compose. I do it that way, because I have many different composed containers running and want to service all of them with one nginx-proxy.

Also do note that if you start nginx-proxy outside the docker-compose file (on its own), you will need to add the docker-compose network to it as well in order for it to work, or make sure your docker-compose forces the default docker network.

**Comment by kajmagnus** [comment]
Hi again, I'd just like to say, that ... if you've run into problems this far, with ED (e.g. maintenance or user-experience or ... anything) — it'd be interesting to know. Perhaps that's something I can fix.  ... And if you only tested ED, but decided to not continue using it, then that'd also be interesting to know about, and the main reasons why-not, if you have time.

(If you have time to answer, then I suppose this GitHub issue isn't the right place — you can instead email me at `kajmagnus3 at gmail ...` or post in [the forum](https://www.effectivediscussions.org/forum/latest). )

Ok and have a nice weekend :- )

**Comment by kajmagnus** [comment]
Hi Gerard, now 1.5 years later, this software is more finished & ready to use, in case you're curious about checking it out again.  I renamed it from EffectiveDiscussions to Talkyard, and the new website is https://www.talkyard.io/

This is still beta software. And I would think that, 1.5 years ago, it caused fairly much confusion & frustration — and most of those things have been fixed now, and other people have started using it for real.

(I'm going to close this whole Issues section in a while, so people will need to contact me via the main support forum instead, at https://www.talkyard.io/forum  )

Have a nice end of the summer, & regards,
Magnus
